<!DOCTYPE html>
{% extends "layout.html" %}

{% block title %}
    Home
{% endblock %}

{% block head %}
    {{ super() }}

    <style>
        .column {
            float: left;
            width: 50%;
        }

        .row {
            margin-left: 10px;
        }

        /* Clear floats after the columns */
        .row:after {
            content: "";
            display: table;
            clear: both;
        }

    </style>

{% endblock %}

{% block content %}
<h2>Welcome to <b><a id="polyoligo" href="https://github.com/MirkoLedda/polyoligo">PolyOligo</a></b> <small>v0.1</small> <h4> <font color="gray">- a suite for designing oligonucleotides on polyploid and complex genomes.</font></h4></h2>
<hr>

    <h4> &darr; Choose your method &darr;</h4>
    <div id="methods">
        <ul style="list-style: none;">
            <li> <b><a href="{{ url_for('pcr') }}">PCR/Sanger</a></b> - Standard PCR or Sanger sequencing primers.
            <li> <b><a href="{{ url_for('kasp') }}">KASP</a></b> - Genotyping by sequencing using the Kompetitive allele specific PCR (KASP) method.
            <li> <b><a href="{{ url_for('caps') }}">CAPS</a></b> - Genotyping by sequencing using the Cleaved Amplified Polymorphic Sequences (CAPS) method.
            <li> <b><a href="{{ url_for('crispr') }}">CRISPR/Cas9</a></b> - Guide RNAs for CRISPR/Cas9 assays.
        </ul>
    </div>
    <br>
    <div id="summary">
        <p>
            <h4><b>Important information - please read before starting</b></h4>
            Complete documentation and a command line interface (CLI) version of this software is available <b><a href="https://github.com/MirkoLedda/polyoligo">HERE</a></b>. The CLI version can be used with any reference genome and VCF file, has more features and is faster because multi-threads enabled.<br><br>

            To avoid overloading the server and transferring large amount of data, we implemented the following limitations here:<br>

            <ol>
                <li>
                    Currently supported genomes are
                    <select name="reference">
                        {% for blastdb_option in blastdb_options %}
                            <option value= "{{blastdb_option}}" {% if blastdb_option=='Fragaria x ananassa (KnappLab)' %} selected="selected"{% endif %}>{{blastdb_option}}</option>"
                        {% endfor %}
                    </select> (more can be added upon <b><a href="https://github.com/MirkoLedda/polyoligo/issues">request</a></b>).
                </li>
                <li>The number of concurrent assay designs is capped.</li>
                <li>The number of threads is limited to 1 per job.</li>
                <li>Mutations across populations other than strawberries cannot be used (i.e. no VCF file upload).</li>
            </ol><br>

            If you find this tool useful, please support us by citing:<br><br>
            <small>Ledda M., Cobo N., Lorant A., Hardigan M.A. and Knapp S.J., PolyOligo: A Bioinformatic Platform for Identifying Target DNA Sequences for the Development of Sub-Genome Specific DNA Markers in Polyploid/Complex Genomes. Poster presented at: Annual Conference of the American Society of Horticultural Sciences; 2019 July 21-25; Las Vegas, NV, USA.</small><br><br>
            For feature requests, reporting bugs or addition of reference genomes/population mutations, please fill <b><a href="https://github.com/MirkoLedda/polyoligo/issues">this form</a></b> or <b><a href="mailto:maledda@ucdavis.edu?Subject=PolyOligo%20support">contact us</a></b>.<br>
        <p>
    </div>

{% endblock %}
